---
title: "ATAC peak readcount"
author: "Romuald Parmentier"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document :
    df_print: kable
    highlight: default
    theme : journal
    number_sections: yes
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
    self_contained : yes
---

# Goal of the script

This script has two parts and it aims to retrieve the number of reads associated to different region of interest and store them in .csv file.\
\
These "readcount_matrix" files are then used in the following script: ATAC_peaks_dynamics.Rmd
\
**Note** : another ReadCount step is performed in the following script : ATAC_differential_accessibility.Rmd \
It's a deliberate choice to not having add this step in this script, because useful intermediate variables are created along this step for the ATAC_differential_accessibility.Rmd script


```{r Dependencies}

library(Rsubread)
library(GenomicRanges)
library(stringr)
library(dplyr)
library(kableExtra)

```

```{r retreive last files crated in the exp}

pic_last_files = function(parent_folder, pattern){
  
  # List dir and get the last created, corresponding to the up-to-date data generated
  dir = list.dirs(path = parent_folder, recursive = F)
  dir = dir[length(dir)]

  files = list.files(
    path = dir,
    pattern = pattern,
    full.names = T)
  
  return(files)
  
}

```

# Readcount on donors intersection at fixed time points

This first part aims to calculate the number of reads associated to the peaks corresponding to a 00h, 05h, 24h or 48h (theses peaks were determined as the intersection region of the three donors at a fixed time point, see "ATAC_peak_annotation" script).\
\
To perform this, this script needs the "peaks_XXh.csv" files and the "XXX_downsampled.bam" files in order to count the reads overlaping with the regions corresponding to peaks.\
\
The output corresponds to a matrix where each line represents a peak, and each of the three columns display the number of reads detected in this peak for one donor.

```{r readcount on donors intersection at fixed time points}

label_correspondance =
  read.delim("/home/romuald/MARS-ATAC/data/broadPeak_label_correspondance.csv")  

peaks_files = pic_last_files(parent_folder = "/home/romuald/MARS-ATAC/exp/ATAC_peak_annotation",
  pattern = "csv")

bam_files = pic_last_files("/home/romuald/MARS-ATAC/exp/ATAC_bam_downsampling",
  pattern = "bam")

times = c("00h", "05h", "24h", "48h")

for(i in 1:length(times)){
  
  # Define peaks (region) where to count reads, according time
  peaks_file_path = peaks_files[which(
    str_detect(
      string = peaks_files, 
      pattern = times[i]))]
  peaks_time = read.csv(peaks_file_path, sep=";")
  peaks_time = peaks_time[,c(1:4,6)] %>%
    dplyr::mutate(GeneID = paste(times[i], X, sep="_peak_") ) %>%
    dplyr::rename(Chr = seqnames, Start = start, End = end, Strand = strand) %>%
    dplyr::select(- X) %>%
    dplyr::select(GeneID, Chr, Start, End, Strand)
  
  # Define bam files associated to the peak (region), according time
  labels_time = as.vector.factor(label_correspondance$Label[label_correspondance$Time %in% times[i]])
  bam_files_time = bam_files[which(str_detect(bam_files, pattern = paste0(labels_time,collapse = "|")))]
  
  readCount <- featureCounts(files = bam_files_time,
                           annot.ext = peaks_time,  
                           isPairedEnd = TRUE,
                           nthreads = 1,
                           countChimericFragments = FALSE,
                           countMultiMappingReads = TRUE)
  
  matrix_count = readCount$counts

  #Create a unique folder for output with th date of the day
  output_dir = format(Sys.time(), "%Y_%m_%d_%Hh_%Mmin")
  dir.create(path = paste0("/home/romuald/MARS-ATAC/exp/ATAC_peaks_readcount/", output_dir))
  
  write.csv2(
    x = matrix_count, 
    file = paste0("/media/romuald/TOSHIBA EXT/Romuald_NGS/MARS-ATAC/exp/ATAC_peaks_readcount/",
      output_dir,
      "/readcount_peaks_", 
      times[i], 
      ".csv"))
  
  print(paste("reacount matrix (.csv file) for peaks at", times[i], "generated"))

  matrix_count[1:100,] %>% 
    kable %>%
    kable_styling("striped", full_width = T) %>% 
    scroll_box()
  
  
}

```

